core/src/xtdb/kv/index_store.clj

```diff
@@ -1060,58 +1081,81 @@
 
             (run! #(cache/evict cav-cache %) (keys tombstones))
 
-            (kv/store transient-kv-store
-                      (for [k (concat bitemp-ks content-ks)]
-                        (MapEntry/create k nil)))
+            (doseq [k (concat bitemp-ks content-ks)]
+              (kv/put-kv kv-tx k nil))
 
-            {:tombstones tombstones})))))
+            {:tombstones tombstones}))
+        (finally
+          (doseq [^Closeable c [snapshot pi ti]]
+            (when c (.close c)))))))
 
   (index-entity-txs [_ entity-txs]
-    (kv/store transient-kv-store
-              (->> (mapcat etx->kvs entity-txs)
-                   (sort-by key mem/buffer-comparator))))
+    (doseq [kv (->> (mapcat etx->kvs entity-txs)
+                    (sort-by key mem/buffer-comparator))]
+      (apply kv/put-kv kv-tx kv)))
 
   (commit-index-tx [_]
-    (with-open [snapshot (kv/new-snapshot transient-kv-store)]
-      (kv/store persistent-kv-store (seq snapshot))))
-
-  (abort-index-tx [_]
-    (with-open [snapshot (kv/new-snapshot transient-kv-store)]
-      (let [{::xt/keys [tx-id tx-time]} tx]
-        ;; we still put the ECAV KVs in so that we can keep track of what we need to evict later
-        ;; the bitemp indices will ensure these are never returned in queries
-        (kv/store persistent-kv-store
-                  (conj (->> (seq snapshot)
-                             (filter (fn [[^DirectBuffer k-buf _v-buf]]
-                                       (= c/ecav-index-id (.getByte k-buf 0)))))
-
-                        (MapEntry/create (encode-failed-tx-id-key-to nil tx-id) mem/empty-buffer)
-                        (MapEntry/create (encode-tx-time-mapping-key-to nil tx-time tx-id) mem/empty-buffer))))))
+    (kv/commit-kv-tx kv-tx))
+
+  (abort-index-tx [_ docs]
+    (when kv-tx
+      (kv/abort-kv-tx kv-tx))
+    (let [{::xt/keys [tx-id tx-time]} tx
+          attr-bufs (->> (into #{} (mapcat keys) (vals docs))
+                         (into {} (map (juxt identity c/->id-buffer))))]
+
+      ;; we still put the ECAV KVs in so that we can keep track of what we need to evict later
+      ;; the bitemp indices will ensure these are never returned in queries
+
+      (->> (for [[content-hash doc] docs
```

Comment:
subjective nit: I tend to only use `->>` when it's a value I'm transforming through the thread, rather than because the value happens to be the last parameter each time - you kind of lose the side-effect here?